ipv8/keyvault/doublesign.py

```diff
@@ -0,0 +1,521 @@
+            res = val ** ((self.p + 1) / 4)
+        elif sw == 5:
+            x = val ** ((self.p + 1) / 4)
+            if x == 1:
+                res = val ** ((self.p + 3) / 8)
+            else:
+                res = (4 * val) ** ((self.p - 5) / 8) * 2 * val
+        else:
+            raise Exception("modsqrt non supported for (p%8)==1")
+        if res.value % 2 == flag:
+            return res
+        return -res
+
+    def inverse(self, value):
+        """
+        Calculates the multiplicative inverse
+        """
+        return _modinv(value.value, self.p)
+
+    def value(self, x):
+        """
+        Converts an integer or FinitField.FieldElement to a value of this FiniteField.
+        """
+        return x if isinstance(x, FieldElement) and x.field == self else FieldElement(self, x)
+
+    def integer(self, x):
+        """
+        Returns a plain integer
+        """
+        return x.value if isinstance(x, FieldElement) else x
+
+    def zero(self):
+        """
+        Returns the additive identity value (a + 0 = a)
+        """
+        return FieldElement(self, 0)
+
+    def one(self):
+        """
+        Returns the multiplicative identity value (a * 1 = a)
+        """
+        return FieldElement(self, 1)
+
+
+class ECPoint(object):
+    """
+    Represents an affine point in the elliptic curve.
+    """
+
+    def __init__(self, curve, x, y):
+        super(ECPoint, self).__init__()
+        self.curve = curve
+        self.x = x
+        self.y = y
+
+    # ECPoint + ECPoint
+    def __add__(self, rhs): return self.curve.add(self, rhs)
+
+    def __sub__(self, rhs): return self.curve.sub(self, rhs)
+
+    # ECPoint * int   or ECPoint * FieldElement
+    def __mul__(self, rhs): return self.curve.mul(self, rhs)
+
+    def __div__(self, rhs): return self.curve.div(self, rhs)
+
+    def __eq__(self, rhs): return self.curve.eq(self, rhs)
+
+    def __ne__(self, rhs): return not self == rhs
+
+    def __str__(self): return "(%s,%s)" % (self.x, self.y)
+
+    def __neg__(self): return self.curve.neg(self)
+
+    def iszero(self):
+        return self.x.iszero() and self.y.iszero()
+
+    def isoncurve(self):
+        return self.curve.isoncurve(self)
+
+
+class EllipticCurve(object):
+    """
+    Elliptic Curve in Montgomery form.
+    """
+
+    def __init__(self, field, a, b):
+        super(EllipticCurve, self).__init__()
+        self.field = field
+        self.a = field.value(a)
+        self.b = field.value(b)
+
+    def add(self, p, q):
+        if p.iszero(): return q
+        if q.iszero(): return p
+
+        if p == q:
+            if p.y == 0:
+                return self.zero()
+            l = (3 * p.x ** 2 + self.a) / (2 * p.y)
+        elif p.x == q.x:
+            return self.zero()
+        else:
+            l = (p.y - q.y) / (p.x - q.x)
+
+        x = l ** 2 - (p.x + q.x)
+        y = l * (p.x - x) - p.y
+        return self.point(x, y)
+
+    def sub(self, lhs, rhs):
+        return lhs + -rhs
```

Comment:
`+ -` is redundant

Reply:
This is the way it should be. Point negation and addition.